beniget/beniget.py

```diff
@@ -971,6 +1085,100 @@ def visit_withitem(self, node):
             self.visit(node.optional_vars)
         return dnode
 
+def _iter_child_nodes_in_fields(node, fields):
+    """
+    Like ast.iter_child_nodes() but ignore fields
+    that are not part of the provided list.
+    """
+    for fieldname, value in ast.iter_fields(node):
+        if fieldname in fields:
+            yield value
+
+def _iter_arguments(node):
+    """
+    Yields all arguments of the given function.
+    """
+    for args in _iter_child_nodes_in_fields(node.args, ('args', 'kwonlyargs', 'posonlyargs')):
+        for a in args: yield a
+    for a in filter(None, _iter_child_nodes_in_fields(node.args, ('vararg', 'kwarg'))):
+        yield a
+
+def _lookup_annotation(name, heads, locals_map):
+    scopes = _get_lookup_scopes(heads)
```

Comment:
Instead of this scope context, couldn't we just walk through the ancestors (that we would need to compute as a prerequisite, but we already have the mechanic to do so in beniget) and for each ancestor statement, check the locals there?